src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs

```diff
@@ -635,10 +635,21 @@ modifyIndices = undefined
 
 -- | Modify each element of an array using the supplied modifier function.
 --
--- /Unimplemented/
-modify :: -- forall m a b. (MonadIO m, Storable a) =>
-    Array a -> (a -> a) -> m ()
-modify = undefined
+-- /Pre-release/
+modify :: forall m a. (MonadIO m, Storable a)
+    => Array a -> (a -> a) -> m ()
+modify Array{..} f = liftIO $ go 0
+
+    where
+
+    go i = do
+        unsafeWithArrayContents arrContents arrStart $ \ptr -> do
+            let elemSize = sizeOf (undefined :: a)
+                elemPtr = ptr `plusPtr` (elemSize * i)
+            when (elemPtr `plusPtr` elemSize <= aEnd) $ do
+                r <- peek elemPtr
+                poke elemPtr (f r)
+                go (i+1)
```

Comment:
Instead of passing the index in the loop we can pass the ptr itself, so that we do not have to multiply with the index each time. We can just keep incrementing the ptr.